### PR TITLE
Change user model

### LINKED ISSUE #104: Extend Illuminate\Foundation\Auth\User
I've found that I am having to handle Auth0 user data and my database models separately which isn't very convenient. Using Eloquent's `User::firstOrNew()` isn't possible using `Auth0User`, but if decide to extend my user object from the default Eloquent object, I lose access to the added Auth0 data, I'd love to have peanut butter *and* chocolate 😄

### PR DESCRIPTION
**Pull request by joshcanhelp**
### Changes

- Add `Auth0UserTrait` 
- More TBD ...

### References

Closes #104 

### Testing

[ ] This change has been tested on the latest version Laravel

### CODE REVIEW

**Comment by Tamrael** [comment]
have you ever thought about converting the Auth0User|Auth0JWTUser from implementing the contract to a trait that would be applied to the model just like the "AuthorizableContract"-trait?

**Comment by joshcanhelp** [comment]
@Tamrael - Thanks for the feedback and apologies for the delay here. 

What's the benefit to going the trait route? I'd like this to be as "Laravel-y" as possible but we don't have a lot of internal knowledge on how the user model works, to be honest. Any guidance you can provide would be appreciated!

**Comment by Tamrael** [comment]
Well as far as I can tell your user models are just using private variables and public functions to format and return what came from the API. The benefit of doing this as a trait is that it could be applied to any model on existing apps and merge eloquent models from the database with and by the data provide from logging in with auth0. Only downside I see is that the eloquent model already uses the constructor and a new app that doesn't need database stored users would need a dummy user model.

I'm out of office till the start of next week but I could write up some example for you once I'm back

**Comment by joshcanhelp** [comment]
@Tamrael - Appreciate the feedback here. I'm open to approaches here and it may be that our idea of the user model is outdated compared to what developers are doing now with Laravel. I've been mainly guided by [this issue](https://github.com/auth0/laravel-auth0/issues/104) which, admittedly, is quite old at this point.

**Comment by aaronflorey** [comment]
I think the more Laravel way would be to have an Auth0 trait on the Users model implementing the needed functions by Auth0, and pass that back from the Auth0 Repository setting the accessToken before returning. This is what I've done in my project.

```
	/**
	 * Get a User from the database using Auth0 profile information
	 *
	 * @param array $userinfo
	 *
	 * @return User
	 */
	public function getUserByUserInfo($userinfo)
	{
		$user = $this->upsertUser($userinfo['profile']);
		$user->accessToken = $userinfo['accessToken'];

		return $user;
		//return new Auth0User($user->getAttributes(), $userinfo['accessToken']);
	}
```

**Comment by joshcanhelp** [comment]
@aaronflorey - Appreciate the feedback here and it looks like you've got some support but I don't think I totally understand what you're proposing. You're saying to convert the `Auth0User` class into a trait?

https://github.com/auth0/laravel-auth0/blob/master/src/Auth0/Login/Auth0User.php

**Comment by joshcanhelp** [comment]
@aaronflorey - Appreciate the feedback here and it looks like you've got some support but I don't think I totally understand what you're proposing. Where does `$this->upsertUser()` come from? And how is the trait invoked in your example?

**Comment by ChalupaBatman69** [comment]
Hey @joshcanhelp,

I'll provide my code that we're currently using. 

https://gist.github.com/ChalupaBatman69/f3bfbae0eb9060349f0dfb490e201559

I have a conditional on there, because we're transitioning from using our old system and the new system.

**Comment by aaronflorey** [comment]
Hey @joshcanhelp,

I'll provide my code that we're currently using.

https://gist.github.com/aaronflorey/d20f27a2b0475d238e10b46de3bc3eb4

I have a conditional on there, because we're transitioning from using our old system and the new system.

**Comment by joshcanhelp** [comment]
@aaronflorey - Thank you for sharing this, this is much more clear! So our methods would be in an Auth0 trait and we'd provide a generic User model, as we do now (maybe using what's in this PR) that uses the trait. Then everyone else could either use that or another model using the Auth0 trait. 

Would that cover the use case you're describing here?

File: src/Auth0/Login/Auth0JWTUser.php [hunk 1]
```diff
@@ -2,99 +2,35 @@
...
+    use Auth0UserTrait;
 
     /**
      * Auth0JWTUser constructor.
      *
-     * @param $userInfo
+     * @param array $userInfo
      */
-    public function __construct($userInfo)
+    public function __construct(array $userInfo)
```
Comment (joshcanhelp) [inline src/Auth0/Login/Auth0JWTUser.php:22]: Breaking change. The userinfo that's generated comes from ID token validation/decoding, which, in the PHP SDK, returns an array. We were converting to an object to pass, then an array to use.

File: src/Auth0/Login/Auth0JWTUser.php [hunk 2]
```diff
@@ -2,99 +2,35 @@
...
-        return $this->userInfo;
-    }
-
-    /**
-     * @return string
-     */
-    public function __toString()
-    {
-        return json_encode($this->userInfo);
+        return $this->userInfo['sub'] ?? null;
```
Comment (joshcanhelp) [inline src/Auth0/Login/Auth0JWTUser.php:34]: I'm not so sure about this but it's, effectively, what was there before. Shouldn't this be `"sub"`? Currently (and now) it's returning the same as `getAuthIdentifier()` in the trait.

File: src/Auth0/Login/Auth0Service.php [hunk 3]
```diff
@@ -162,12 +162,13 @@ public function rememberUser($value = null)
...
      *
-     * @return object
+     * @return array
+     *
      * @throws \Auth0\SDK\Exception\InvalidTokenException
      */
     public function decodeJWT($encUser, array $verifierOptions = [])
     {
-        $this->apiuser = (object) $this->auth0->decodeIdToken($encUser, $verifierOptions);
+        $this->apiuser = $this->auth0->decodeIdToken($encUser, $verifierOptions);
```
Comment (joshcanhelp) [inline src/Auth0/Login/Auth0Service.php:171]: Breaking change. Again, we were converting to an object then back to an array.

File: src/Auth0/Login/Contract/Auth0UserRepository.php [hunk 4]
```diff
@@ -2,26 +2,28 @@
...
 {
     /**
-     * @param stdClass $jwt with the data provided in the JWT
+     * @param array $jwt with the data provided in the JWT
      *
-     * @return \Illuminate\Contracts\Auth\Authenticatable
+     * @return Authenticatable
      */
-    public function getUserByDecodedJWT($jwt);
+    public function getUserByDecodedJWT(array $jwt);
```
Comment (joshcanhelp) [inline src/Auth0/Login/Contract/Auth0UserRepository.php:14]: Breaking change, same as the others above.

File: src/Auth0/Login/Contract/Auth0UserRepository.php [hunk 5]
```diff
@@ -2,26 +2,28 @@
...
 
     /**
      * @param array $userInfo representing the user profile and user accessToken
      *
-     * @return \Illuminate\Contracts\Auth\Authenticatable
+     * @return Authenticatable
      */
-    public function getUserByUserInfo($userInfo);
+    public function getUserByUserInfo(array $userInfo);
```
Comment (joshcanhelp) [inline src/Auth0/Login/Contract/Auth0UserRepository.php:21]: Not breaking, was expecting an array, just making it official

File: src/Auth0/Login/Repository/Auth0UserRepository.php [hunk 6]
```diff
@@ -9,11 +9,11 @@
 class Auth0UserRepository implements Auth0UserRepositoryContract
 {
     /**
-     * @param \Auth0\Login\Contract\stdClass $jwt
+     * @param array $jwt Verified and decoded JWT.
```
Comment (joshcanhelp) [inline src/Auth0/Login/Repository/Auth0UserRepository.php:12]: Breaking change, see above.

File: src/Auth0/Login/Repository/Auth0UserRepository.php [hunk 7]
```diff
@@ -9,11 +9,11 @@
...
      *
      * @return Auth0JWTUser
      */
-    public function getUserByDecodedJWT($jwt)
+    public function getUserByDecodedJWT(array $jwt) : Auth0JWTUser
```
Comment (joshcanhelp) [inline src/Auth0/Login/Repository/Auth0UserRepository.php:16]: Breaking change, see above.

File: src/Auth0/Login/Traits/Auth0UserTrait.php [hunk 8]
```diff
@@ -0,0 +1,112 @@
+<?php
```
Comment (joshcanhelp) [inline src/Auth0/Login/Traits/Auth0UserTrait.php:1]: Here's the trait we were discussing. Is this enough to be helpful on other user models?

File: src/Auth0/Login/Traits/Auth0UserTrait.php [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * Get id field name.
+     *
+     * @return string
+     */
+    public function getAuthIdentifierName() : string
+    {
+        return 'id';
```
Comment (joshcanhelp) [inline src/Auth0/Login/Traits/Auth0UserTrait.php:34]: This is confusing to me ... should this continue to be `"id"` as it was before? Seems like it should be `"sub"` but maybe I'm not understanding the purpose.

File: src/Auth0/Login/Traits/Auth0UserTrait.php [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * Get the password for the user.
+     *
+     * @return string|null
+     */
+    public function getAuthPassword() : ?string
```
Comment (joshcanhelp) [inline src/Auth0/Login/Traits/Auth0UserTrait.php:42]: This is now a part of `Auth0JWTUser`. Before that returned nothing/void and now will always return null.

**Comment by joshcanhelp** [comment]
@aaronflorey @Tamrael @alexbuckham - Appreciate your guidance on this so far. 

I put together what I think is a good trait-based strategy for handling users. `Auth0User` and `Auth0JWTUser` remain generally unchanged and still produce an `Authenticatable` but not the trait can be attached to, say, the Eloquent user model. I also adjusted many places where the types used were unhelpful or undefined. 

The only real breaking change here is switching what `Auth0JWTUser` expects in the constructor from an object to an array (see the comments for the rationale). Most of this could be done in a minor release but improving the types feels like a solid improvement (assuming, of course, that they are correct :D).

I would love your (or anyone watching this PR) feedback on whether these changes are helpful or if I missed anything. If we can't get consensus on this change, we'll keep it out of the pending major release and see what we can do minor release when we can iron out the details. 

cc: @jimmyjames

**Comment by craigpotter** [comment]
I think adding the function to attach a trait to an existing model is the right way forward. I'm currently rewriting one of our applications from Laravel 5.8 to 7 and decided to use the latest version of this package. 

Although most problems I had looked to be solved, as a Laravel user, I would expect `auth()->user()` to return the current User eloquent model. 

The trait also makes it easier if I want multiple Auth schemes in the same application. One for customers and one for admins. I can just attach the trait to the customer model and apply the correct middleware.

**Comment by joshcanhelp** [comment]
@craigpotter - Appreciate the feedback here. Does this PR satisfy your needs in the scenario you described?

**Comment by bryanjamesmiller** [comment]
@joshcanhelp Thanks for getting back to me on that separate GitHub issue - I appreciate it!  This branch would also seem to serve my purposes, as we are considering switching out of our Lock JS / PHP hybrid approach to the pure PHP approach - however, as things currently are we also have the issue with `auth()->user()` returning an instance of `Auth0User` when we are already using `auth()->user()` as an eloquent model throughout the app.  I was hoping putting this trait on my User model would make everything right, but I got this error:

Screenshot at Apr 22 22-24-11
 
Hope this helps you out and if I'm missing something, please let me know.  Thanks again!

**Comment by joshcanhelp** [comment]
@bryanjamesmiller - Appreciate the feedback here. Can you provide a code sample for the User model that's causing this error?

**Comment by mttpg** [comment]
I'm really pleased to see that these changes are coming. We're keen to integrate Auth0 with our existing web applications, but being able to keep our current user models is a very important factor in deciding whether we can go ahead.

I've tried the latest version from the `change-user-model` branch but keep hitting problems, any tips on how to get it working?

**Comment by joshcanhelp** [comment]
> I've tried the latest version from the `change-user-model` branch but keep hitting problems, any tips on how to get it working?

Can you be more specific about the problems you're having?

**Comment by mttpg** [comment]
> Can you be more specific about the problems you're having?

I spent a long time having compatibility issues between the interfaces, but the new contracts and traits seem to have solved that. The new error is this one - **Can't initialize a new session while there is one active session already** - which happens during the login process (using a clean browser, no old cookies, etc). 

It seems to be related to my custom user repository class (per this comment https://github.com/auth0/laravel-auth0/pull/165#issuecomment-600913694), which I have modified to return my `App\User` object instead of the `Auth0User` object (as mentioned before I need to use the existing User model). The `App\User` class implements `Auth0UserTrait`. 

I understand this is still in development and that the documentation hasn't been updated yet, so I've probably done the wrong thing somewhere. It would be really helpful if you could provide some clarity on how the Auth0 stuff is supposed to integrate with the Laravel user model.

Many thanks

**Comment by craigpotter** [comment]
@joshcanhelp Any further thoughts or progress on this one ? 

I am hoping to test it properly over the weekend if I can as I am still using v5.7 in my projects so would be keen to make the jump.

**Comment by bkilshaw** [comment]
@joshcanhelp I would be happy to test this out if you are able to provide some quick notes on how to migrate to this version from the last.

**Comment by jontas** [comment]
@joshcanhelp also watching this hoping it is merged soon!  Happy to help test if that is useful.

**Comment by evansims** [comment]
Hi everyone, apologies for the delay on this. I've just joined the team this week, and will be catching up on some of these lingering issues shortly, once I've had time to investigate and familiarize myself with everything. Hang in there a bit longer, we'll get this sorted.

**Comment by kingsloi** [comment]
@evansims that's great news!

**Comment by bkilshaw** [comment]
@evansims Great news Evan! Do you happen to have any updates on this? I've been waiting for Auth0 to allow me to use my regular eloquent User model for ages. Super high priority for us.

**Comment by bkilshaw** [comment]
@Tamrael you seem to have a pretty good grasp on implementing and customizing this package. I can't find a way to reach out to you directly, if you're available for hire to help customize our implementation please shoot me a note at brad [at] nivel dot tech

**Comment by mttpg** [comment]
> @evansims Great news Evan! Do you happen to have any updates on this? I've been waiting for Auth0 to allow me to use my regular eloquent User model for ages. Super high priority for us.

This is high priority for us too. Would appreciate an update on what's happening - thanks!

**Comment by evansims** [comment]
@mttpg @bkilshaw Totally understandable, and I recognize this being a high priority for you all. As I'm still catching up and familiarizing myself with these code bases I'm not able to give you a hard timeline yet, but this is a priority item on my list as well. Hang in there a bit longer.

**Comment by Tamrael** [comment]
> @Tamrael you seem to have a pretty good grasp on implementing and customizing this package. I can't find a way to reach out to you directly, if you're available for hire to help customize our implementation please shoot me a note at brad [at] nivel dot tech

Sorry for being late to answer. Thank you for the offer but I'll have to decline.

However that doesn't mean I won't try to help.
I made a [simple gist](https://gist.github.com/Tamrael/48ddc55a30b2170ec6303bf7db6b0255) that should get you started on what I did when we implemented auth0 for our laravel project with existing users. If you have questions just comment on the gist so we can document this for the community.

**Comment by mttpg** [comment]
> @mttpg @bkilshaw Totally understandable, and I recognize this being a high priority for you all. As I'm still catching up and familiarizing myself with these code bases I'm not able to give you a hard timeline yet, but this is a priority item on my list as well. Hang in there a bit longer.

Thanks @evansims. Any news on when these updates might be available?

**Comment by evansims** [comment]
> Thanks @evansims. Any news on when these updates might be available?

Not able to provide a timeline right now, but this is still very much on my radar. My focus has been on getting PHP 8 support shipped throughout our PHP ecosystem lately, but after that I'll be able to turn my attention back to this. Thanks for your patience.

**Comment by evansims** [comment]
👋 Going to close this PR for now. I'm giving the Laravel plugin a rewrite for the next major which will greatly improve this experience, much in the same fashion as what I'm currently wrapping up for [the underlying PHP SDK](https://github.com/auth0/auth0-PHP). The efforts and feedback on this PR are appreciated and will inform my efforts on the next version. 👍 Thank you everyone

**Pull request closed without merging**